### PR TITLE
[WRANGLER-60] Extract Fields directives in dataprep

### PR DESCRIPTION
**Pull request by ajainarayanan**
- Adds ExtractFields directive to point and click directives

### TODO:
- Adds regex for `Date pattern`
- Some of the regex are not working as expected. Need to fix them
- ~Dropdown menu is not as per the JIRA. Need to add examples the way its mentioned in the JIRA.~  DONE

### CODE REVIEW

File: cdap-ui/app/cdap/components/DataPrep/Directives/ExtractFields/UsingPatternsModal/index.js [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+      startAndEndPattern: {
+        start: '',
+        end: ''
+      },
+      nDigitPattern: '',
+      error: '',
+      showEditPatternLabel: false,
+      showEditPatternTxtBox: false
+    };
+    this.patterns = [
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/DataPrep/Directives/ExtractFields/UsingPatternsModal/index.js:48]: for simplicity in updating this pattern in the future, can we perhaps declare a `const` of a map of pattern name and the regex value? Then we don't need to add the individual labels/examples. If we follow a pattern in the i18n naming to patternName as the key, then the i18n can just be added inline

File: cdap-ui/app/cdap/components/DataPrep/Directives/ExtractFields/UsingPatternsModal/index.js [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+
+import React, {Component, PropTypes} from 'react';
+import {Modal, ModalHeader, ModalBody, ModalFooter} from 'reactstrap';
+import T from 'i18n-react';
+import isNil from 'lodash/isNil';
+import isEmpty from 'lodash/isEmpty';
+import {Input, DropdownToggle, DropdownMenu, DropdownItem} from 'reactstrap';
+import {execute} from 'components/DataPrep/store/DataPrepActionCreator';
+const PREFIX = 'features.DataPrep.Directives.ExtractFields.UsingPatterns';
+import MouseTrap from 'mousetrap';
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/DataPrep/Directives/ExtractFields/UsingPatternsModal/index.js]: `Mousetrap`?

**Comment by elfenheart** [comment]
Is there going to be more options on the submenu of Extract Fields? Seems weird to only have 1 option that opens a modal

**Comment by ajainarayanan** [comment]
@elfenheart yeah. For now we only have one suboption but we will have more in the near future.

File: cdap-ui/app/cdap/components/DataPrep/Directives/ExtractFields/UsingPatternsModal/index.js (hunk 1 above)
Comment (ajainarayanan) [reply cdap-ui/app/cdap/components/DataPrep/Directives/ExtractFields/UsingPatternsModal/index.js:48]: Talked to @elfenheart. Right now the patterns have mostly just simple string values but some also have `getRegex` function which needs to be evaluated based on user input which will update the actual pattern. For now lets have it as plain array and tomorrow when we add more regex patterns and if it becomes a hassle will extract this out to make it more easier.

**Comment by tonybach** [comment]
@ajainarayanan For patterns where the user has to input text (N digits number, Start/End, and Custom), can you set autofocus on the first input box of that modal? Just to be consistent with other directives we have.

File: cdap-ui/app/cdap/components/DataPrep/Directives/ExtractFields/UsingPatternsModal/index.js [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+
+import React, {Component, PropTypes} from 'react';
+import {Modal, ModalHeader, ModalBody, ModalFooter} from 'reactstrap';
+import T from 'i18n-react';
+import isNil from 'lodash/isNil';
+import isEmpty from 'lodash/isEmpty';
+import {Input, DropdownToggle, DropdownMenu, DropdownItem} from 'reactstrap';
+import {execute} from 'components/DataPrep/store/DataPrepActionCreator';
+const PREFIX = 'features.DataPrep.Directives.ExtractFields.UsingPatterns';
+import Mousetrap from 'Mousetrap';
```
Comment (elfenheart) [inline cdap-ui/app/cdap/components/DataPrep/Directives/ExtractFields/UsingPatternsModal/index.js]: the from is lowercase:

`import Mousetrap from 'mousetrap'`

**Comment by tonybach** [comment]
Code LGTM 👍

**Comment by bdmogal** [comment]
LGTM. We need to add keyboard interactions for the dropdown. Can you open a jira for that?

**Comment by ajainarayanan** [comment]
@bdmogal https://issues.cask.co/browse/CDAP-9458

**Pull request merged**